File: coderbyte/tree-constructor.js
function TreeConstructor(strArr) {
    let newStrArr = [];
    strArr.forEach(i => {
        // '(1,2)' -> [1,2]
        let a = parseInt(i.split(',')[0].slice(1));
        let b = parseInt(i.split(',')[1].slice(0, -1));
        newStrArr.push([a, b]);
    })
    strArr = newStrArr;

    const children = [];
    const parents = [];
    strArr.forEach(i => {
        children.push(i[0]);
        parents.push(i[1]);
    })
    function oneRoot() {
        const roots = [];
        parents.forEach(i => {
            if (!children.includes(i)) roots.push(i);
        })
        return new Set(roots).size === 1;
    }

    function oneParent() {
        return children
            .map(i => children.filter(x => x === i).length)
            .every(k => k === 1);
    }

    function atLeastTwoChildren() {
        return parents
            .map(i => parents.filter(x => x === i).length)
            .every(k => k <= 2);
    }

    if (!oneRoot()) {
        console.log('not one root');
        return false;
    }
    else if (!oneParent()) {
        console.log('not one parent');
        return false;
    }
    else if (!atLeastTwoChildren()) {
        console.log('more than two children');
        return false;
    }
    else {
        return true;
    }
}

console.log(TreeConstructor(["(10,20)"]));

File: coderbyte/tree-constructor.py
# Input: ["(1,2)", "(2,4)", "(5,7)", "(7,2)", "(9,5)"]
# Output: true
#
# Input: ["(1,2)", "(3,2)", "(2,12)", "(5,2)"]
# Output: false

def TreeConstructor(strArr: list) -> bool:
    newStrArr = []
    for i in strArr:
        a = int(i.split(',')[0][1:])
        b = int(i.split(',')[1][:-1])
        newStrArr.append((a, b))

    strArr = newStrArr
    children = []
    parents = []
    for i in strArr:
        children.append(i[0])
        parents.append(i[1])

    def oneRoot():
        'Checks if tree has one root node'
        roots = []
        for i in parents:
            if i not in children:
                roots.append(i)
        return len(set(roots)) == 1

    def oneParent():
        'Checks if every child node has only one parent'
        return all(children.count(i) == 1 for i in children)

    def atLeastTwoChildren():
        'Checks if every parent has at least two children'
        return all(parents.count(i) <= 2 for i in parents)

    if not oneRoot():
        # print('not one root')
        strArr = False

    elif not oneParent():
        # print('not one parent')
        strArr = False

    elif not atLeastTwoChildren():
        # print('more than two children')
        strArr = False

    else:
        strArr = True

    return strArr

print(TreeConstructor(["(1,2)", "(2,4)", "(7,2)"]))
